### PR TITLE
Procedural Level Builder

### PR DESCRIPTION
**Pull request by Oggal**
Want to do a solid Code Review over this!

Please ask any questions you have on how this script works!

Safe, and Dangerous Prefabs hold the clouds to be spawned, with thunder in Dangerous and Walk Clouds in safe.

Background is unused, added when thinking about adding clouds to the background procedurally...

Max height is the point the animation curve ends at, after that point the game doesn't get more difficult, but still continues

min and max step is distance on y between clouds,
maxShift is the max distance left and right between clouds

lastPlaced is the last Cloud placed, assigned first by hand as the highest cloud placed by hand.

levelWidth is width of the scene, it gets set properly at start.

lastly, the List<GameObject> clouds is a list of procedural clouds.

Going over this, I see at least one thing I need to fix, (what happens when the player loses and we need to reset the level?)
if you can find anything else wrong let me know!

### CODE REVIEW

File: Assets/Scripts/LevelBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        GetWidth();
+        lastGenerated = lastPlaced;
+    }
+
+    // Update is called once per frame
+    void Update()
+    {   
+        if (lastGenerated.transform.position.y < Camera.main.transform.position.y + Camera.main.orthographicSize + maxStep)
+            clouds.Add(PlaceNewCloud());
+        for(int i = clouds.Count - 1; i >= 0; i--)
```
Comment (teylorchapman) [inline Assets/Scripts/LevelBuilder.cs:37]: For a second there, I was confused as to why you were doing the (""; ""; i--) but as i read further down I got it, what you're doing here is removing the cloud after the camera has panned up.

File: Assets/Scripts/LevelBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        float chance = Random.value;
+
+        float index = (SafeChance.Evaluate(transform.position.y / MaxHeight) + Random.value);
+
+        if (chance > SafeChance.Evaluate(transform.position.y / MaxHeight))
+            return SafePrefabs[Mathf.RoundToInt(index * SafePrefabs.Length ) % SafePrefabs.Length];
+        else
+            return DangerPrefabs[Mathf.RoundToInt(index * DangerPrefabs.Length ) % DangerPrefabs.Length];
+    }
```
Comment (teylorchapman) [inline Assets/Scripts/LevelBuilder.cs:63]: So as I understand here, when it's generating the next set of platforms, the SafeChance means that it is a safe platform and DangerPrefabs is definitely not a safe platform.

File: Assets/Scripts/LevelBuilder.cs (hunk 1 above)
Comment (Oggal) [reply Assets/Scripts/LevelBuilder.cs:37]: Yes! And looping backward through the list means we can remove items from it without modifying the index of remaining items

File: Assets/Scripts/LevelBuilder.cs (hunk 2 above)
Comment (Oggal) [reply Assets/Scripts/LevelBuilder.cs:63]: Exactly

**Review by teylorchapman — APPROVED** [review]

**Pull request merged**